README
Flickerize is a collection of scripts to sign e resize photo to upload then on some photo sharing service!
Optimized for flickr.com

## Dependencies ##
	exiv2
	imagemagik

If doesn't exist create "../toUpload" folder and "upload.txt" file

after configured "upload.txt" file

	where upload.txt file structure is:
	#<photo path>			<sign color>		<sign position>		<photo title>
	2010/03/23/DSC_0015.JPG		W			BL			Photo-Title
	2010/03/23/DSC_0020.JPG		B			BR			Photo-Title
	2010/03/23/DSC_0031.JPG		W			BR			Photo-Title
	2010/03/23/DSC_0059.JPG		W			BL			Photo-Title

	where:
	<photo path> is relative photo path from flickerize.sh file
	<sign color> can be: "B" or "W"
	<sign position> can be: "BL" or "BR"
	<photo title> file destination name (if not specified orig file name is used)

from flickerize.sh script folder use:

	cat upload.txt | while read num color pos;do ./toFlickr.sh $num $color $pos;done

or run

	./flickerize.sh




Changelog
2010-06-18  Emanuele Grande <caccolangrifata@gmail.com>

  * Changelog: add e popolate changelog file

  * addSignature.sh, toFlickr.sh, upload.txt: Insert debug info

2010-05-30  Emanuele Grande <caccolangrifata@gmail.com>

  * README, addSignature.sh, flickerize.sh, toFlickr.sh: Add possibility to add
  photo title and change output info structure

2010-05-19  Emanuele Grande <caccolangrifata@gmail.com>

  * toFlickr.sh: add a TODO comment

  * addSignature.sh, toFlickr.sh: add a todo e remove error on std output

  * addSignature.sh, firma2_b.png: Update sign method, now sign is
  proportionated to image size!

  * upload.txt: delete useless fiel

  * upload.txt: signed some photo

2010-05-10  Emanuele Grande <caccolangrifata@gmail.com>

  * README, upload.txt: some adjust to README file and add upload.txt file

  * upload.txt: Del upload.txt file

2010-05-09  Emanuele Grande <caccolangrifata@gmail.com>

  * addSignature.sh: correct some typo and adjust some variable

  * addSignature.sh: adjust some part of the code

  * addSignature.sh: Add dimension auto reconize!

2010-05-08  Emanuele Grande <caccolangrifata@gmail.com>

  * addSignature.sh: initiate autoreconize photo size.

  * addSignature.sh: add TODO list

2010-04-26  Emanuele Grande <caccolangrifata@gmail.com>

  * addSignature_0.2.sh, firma.xcf, firma2.xcf, firma_b.xcf,
  rename_adobeRGB.sh, sign/firma.xcf, sign/firma2.xcf, sign/firma_b.xcf,
  toFlickr_0.2.sh: adjust file a dir structure/organization.

  * README: Adjust!

  * addSignature.sh, addSignature_0.2.sh, toFlickr.sh, toFlickr_0.2.sh,
  upload.txt: edit some comment on 0.1 and 0.2 version of the script


toFlickr.sh
#!/bin/sh
#Prepare photo to upload.

#Source file
ORIG="$1"
#Signature color
COLOR="$2"
#Signature position
POS="$3"
#Photo Title
#if [ "$4" = "" ]
#then
#	TITLE=${ORIG##*/}
#	TITLE=${TITLE%*.JPG}
#	TITLE="$TITLE""_sign.JPG"
#else
	TITLE="$4.jpg"
#fi
#DebugInfo
DEBUG=1

usage ()
{
	echo "usage -- $0 <photo sorce> <sign color> <sign pos> <photo title>";
	echo "\t<photo sorce>\t\tphoto source file"
	echo "\t<sign color>\t\tSet the color of the sign, default is white. Can be:"
	echo "\t\t\t\t\tW -- White sign"
	echo "\t\t\t\t\tB -- Black sign"
	echo "\t<sign pos>\t\tWhere to put the sign, default is BR. Can be:"
	echo "\t\t\t\t\tBR -- (Bottom Right)"
	echo "\t\t\t\t\tBL -- (Bottom Left)"
	echo "\t<photo title>\t\tPhoto title"
}

#Sign photo
if [ "$ORIG" = "" ]
then
	usage;
	return $OK;
fi
if [ "$ORIG" = "-help" ]
then
	usage;
	return $OK
fi

echo "Processing photo $ORIG"

./addSignature.sh $ORIG $COLOR $POS
#End Sign photo


#Resize photo
#Source signed file
SOURCE=${ORIG%*.JPG}
SOURCE=$(echo "$SOURCE""_sign.JPG")
#echo "\t\t$SOURCE"
#Output file
## add "_sign_resized"
DEST=${ORIG%*.JPG}
DEST=$(echo "$DEST""_sign_resized.JPG")
#Take photo name
RDEST=${DEST##*/}
PFOLDER=${DEST%*/*.JPG};

echo "\tCoping and resizing photo..."
if [ -z $4 ]
then
	TITLE=${ORIG##*/}
	TITLE=${TITLE%*.JPG}
	TITLE="$TITLE""_sign.JPG"
else
	DEST=$PFOLDER/$TITLE;
	RDEST=$TITLE;
	echo "SI"
	if [ $DEBUG -eq 1 ]
	then
		echo "\t#### DEBUG INFO ####\n\tPFOLDER:$PFOLDER\tDEST:$DEST\tRDEST:$RDEST"
	fi
fi

if [ $DEBUG -eq 1 ]
then
	echo "\t#### DEBUG INFO ####\n\tTITLE: $TITLE"
	echo "\t#### DEBUG INFO ####\n\tSOURCE: $SOURCE"
	echo "\t#### DEBUG INFO ####\n\tDEST: $DEST"
	echo "\t#### DEBUG INFO ####\n\tRDEST: $RDEST"
fi

cp $SOURCE $DEST
mogrify -geometry 1280 $DEST
#End Resize photo

#Move resized photo to toUpload folder
mv $DEST ../toUpload/$RDEST
echo "\tFile stored to ../toUpload/$RDEST"
#backup photo to signed/$TITLE
mkdir $PFOLDER/signed 2> /dev/null
mv $SOURCE $PFOLDER/signed/$TITLE
echo "Processing photo $ORIG Done!"

exit



toFlickr_0.2.sh
#!/bin/sh
#Prepare photo to upload.

#Source file
ORIG="$1"
#Signature color
COLOR="$2"
#Signature position
POS="$3"
#TODO
#Photo Orientation
ORI="$4"

usage ()
{
	./addSignature.sh -help
}

#Sign photo
if [ "$ORIG" = "-help" ];then
	usage;
	return $OK
fi

#TODO
#Check if file exist
#if [ -f $ORIG ];then
#	echo "File seen to found";
#	exit;
#else
#	echo "File not found";
#	exit;
#fi

echo "Processing photo $ORIG"

./addSignature.sh $ORIG $COLOR $POS
#End Sign photo

#Resize photo
#Source file
SOURCE=${ORIG%*.JPG}
SOURCE=$(echo "$SOURCE""_sign.JPG")
#Output file
DEST=${ORIG%*.JPG}
DEST=$(echo "$DEST""_sign_resized.JPG")

echo "Coping and resizing photo..."
cp $SOURCE $DEST
mogrify -geometry 1280 $DEST
#End Resize photo

#Move resized photo to toUpload folder
#if [ [ -f "toUpload" ] ];then
#	mkdir toUplad
#fi
RDEST=${DEST##*/}
mv $DEST toUpload/$RDEST
echo "File stored to toUpload/$RDEST"
#End move
echo "Done!"

exit



addSignature_0.2.sh
#!/bin/sh
#Add signature to the photo.

#TODO
#1.firma foto						- DONE
#2.ridimensiona foto				- DONE
#3.copia nella cartella toUpload	- DONE
#4.check if file exist				- TODO
#5.generalize on file renaming		- TODO
#6.photo orientation				- TODO
#7.auto reconize photo size			- TODO

#Input vars
#Source file
ORIG="$1"

#Output file
DEST=${ORIG%*.JPG}
DEST=$(echo "$DEST""_sign.JPG");

#Error code
OK=0		#everuthing is ok
POS_ERR=1	#return 1 if sign position is no specified

#Sign file
WSIGN="firma2_w.png"
BSIGN="firma2_b.png"
SIGN=$WSIGN

#TODO Reconize photo orientation, orizontal or vertical photo...
#"sinistra - basso" = vertical photo
#"alto - sinistra" = orizontal photo
#where put the sign
#PHOTOO=$( exif -t 0x0112 $ORIG | grep -i value )
#if [ "PHOTOO" = "Value: sinistra - basso" ];then
	#Vertical photo
	
#else
#	if [ "PHOTOO" = "Value: sinistra - alto" ];then
	#Vertical photo
#	fi
#fi

#ORIZONTAL PHOTO
#BOTTOM RIGHT
BR="+3260+2620"
#BOTTOM LEFT
BL="+20+2620"
#VERTICAL PHOTO
#BOTTOM RIGHT
VBR="+2620+3260"
#BOTTOM LEFT
VBL="+2620+20"

#Default is BR
SIGNPOS="$BR"

usage ()
{
	echo "usage -- $0 <photo sorce> <sign color> <sign pos> <photo orien>";
	echo "\t<photo sorce>\t\tphoto source file"
	echo "\t<sign color>\t\tSet the color of the sign, default is white. Can be:"
	echo "\t\t\t\t\tW -- White sign"
	echo "\t\t\t\t\tB -- Black sign"
	echo "\t<sign pos>\t\tWhere to put the sign, default is BR. Can be:"
	echo "\t\t\t\t\tBR -- (Bottom Right)"
	echo "\t\t\t\t\tBL -- (Bottom Left)"
	echo "\t<photo orien>\t\tPhoto orientation, can be O (orizontal) or V (vertical)"
}

#if [ [$ORIG = "-help"] || [$ORIG = ""]];then
if [ "$ORIG" = "-help" ];then
	usage;
	return $OK
fi

#Signature color
if [ "$2" = "" ];then
	echo "Signature color will be white!"
else
	if [ "$2" = "W" ];then
		SIGN="$WSIGN"
		echo "Signature color will white!"
	fi
	if [ "$2" = "B" ];then
		SIGN="$BSIGN"
		echo "Signature color will black!"
	fi
fi

#Signature position
if [ "$4" = "O"];then
	echo "Photo orientation is orizontal!"
	if [ "$3" = "" ];then
		echo "Signature position will be bottom right!"
	#	usage;
	#	return $POS_ERR;
	else
		if [ "$3" = "BR" ];then
			SIGNPOS="$BR"
			echo "Signature position will be bottom right!"
		fi
		if [ "$3" = "BL" ];then
			SIGNPOS="$BL"
			echo "Signature position will be bottom left!"
		fi
	fi
else
	if [ "$4" = "V" ];then
		echo "Photo orientation is vertical!"
		if [ "$3" = "BR" ];then
			SIGNPOS="$VBR"
			echo "Signature position will be bottom right!"
		fi
		if [ "$3" = "BL" ];then
			SIGNPOS="$VBL"
			echo "Signature position will be bottom left!"
		fi
	fi
fi

composite -geometry $SIGNPOS $SIGN $ORIG $DEST;

echo "Photo signed output is $DEST";

exit


addSignature.sh
#!/bin/sh
#Add signature to the photo.

#TODO
#1.sign photo					- DONE
#2.check if file exist				- TODO
#3.generalize on file renaming			- TODO
#4.photo orientation (oriz or vert)		- TODO
#5.auto reconize photo size			- DONE
#6.resize sign to check photo size		- DONE
#7.Control if more then 1 photo is passed	- TODO

#Input vars
#Source file
ORIG="$1";
SC="$2";
SP="$3";

#Output file
DEST=${ORIG%*.JPG};
DEST=$(echo "$DEST""_sign.JPG");

#Error code
OK=0;		#everuthing is ok
POS_ERR=1;	#return 1 if sign position is no specified

#Sign file
WSIGN="firma2_w.png";	#white signature
BSIGN="firma2_b.png";	#black signature
SIGN=$WSIGN;

usage ()
{
	echo "usage -- $0 <photo sorce> <sign color> <sign pos>";
	echo "\t<photo sorce>\t\tphoto source file";
	echo "\t<sign color>\t\tSet the color of the sign, default is white. Can be:";
	echo "\t\t\t\t\tW -- White sign";
	echo "\t\t\t\t\tB -- Black sign";
	echo "\t<sign pos>\t\tWhere to put the sign, default is BR. Can be:";
	echo "\t\t\t\t\tBR -- (Bottom Right)";
	echo "\t\t\t\t\tBL -- (Bottom Left)";
}

#TODO add OR statement
if [ "$ORIG" = "" ];then
	usage;
	return $OK;
fi
if [ "$ORIG" = "-help" ];then
	usage;
	return $OK;
fi

#Calculate image size
SIZE=$(exiv2 $ORIG | grep -i "image size") 2> /dev/null;
SIZE=${SIZE#*:*};
HEIGHT=${SIZE#*x*};
WIDTH=${SIZE%*x*};
#echo "Image:" $WIDTH "x" $HEIGHT

#Resize sign file
: $((PERCENT = $WIDTH * 100 / 4288 ))
#echo "Percentuale di ridimensionamento:" $PERCENT"%"

#TODO Adjust this "if"
#Signature color
if [ "$2" = "" ];then
	echo "\tSignature color will be white!";
else
	if [ "$2" = "W" ];then
		SIGN="$WSIGN";
		echo "\tSignature color will white!";
	fi
	if [ "$2" = "B" ];then
		SIGN="$BSIGN";
		echo "\tSignature color will black!";
	fi
fi

#Prepare sign file
cp "$SIGN" "$SIGN""_tmp"
mogrify -resize "$PERCENT""%" "$SIGN""_tmp"
SIGN="$SIGN""_tmp"
#echo $SIGN

#Calculate sign size for better positioning
SIGNSIZE=$(exiv2 $SIGN | grep -i "image size") 2> /dev/null;
SIGNSIZE=${SIGNSIZE#*:*};
SHEIGHT=${SIGNSIZE#*x*};
SWIDTH=${SIGNSIZE%*x*};
#echo "Sign:" $SWIDTH"x"$SHEIGHT

#Calculate sign position #TODO Change number with $SHEIGHT $SWIDTH vars
: $((XPOS = $WIDTH - $SWIDTH - 20)); #1020 = Photo xSize - ($SIGN Width + 20)
: $((YPOS = $HEIGHT - $SHEIGHT - 20)); #220 = Photo ySize - ($SIGN Height + 20)

#ORIZONTAL PHOTO
#BOTTOM RIGHT
BR="+$XPOS+$YPOS";
#BOTTOM LEFT
BL="+20+$YPOS";
#VERTICAL PHOTO
#BOTTOM RIGHT
VBR="+2620+3260";
#BOTTOM LEFT
VBL="+2620+20";

#echo "Image size:$WIDTH x $HEIGHT", "Sign Pos: BR_$BR BL_$BL"

#Default is BR
SIGNPOS="$BR";

#Signature position
if [ "$3" = "" ];then
	echo "\tSignature position will be bottom right!";
else
	if [ "$3" = "BR" ];then
		SIGNPOS="$BR";
		echo "\tSignature position will be bottom right!";
	fi
	if [ "$3" = "BL" ];then
		SIGNPOS="$BL";
		echo "\tSignature position will be bottom left!";
	fi
fi

#cp $ORIG $DEST
composite -geometry $SIGNPOS $SIGN $ORIG $DEST;
rm $SIGN

echo "\tPhoto signed output is $DEST";

exit
